### PR TITLE
Recommend deterministic=False for RT-DETR training

### PR DESCRIPTION
**Pull request by artest08**
F.grid_sample has no deterministic CUDA backward, so deterministic=True costs speed without making runs reproducible.

RT-DETR cannot train deterministically on CUDA, but `default.yaml` ships `deterministic: True`, so every RT-DETR training pays the cost of deterministic algorithms while still producing non-reproducible results. This adds a tip next to the training example on the RT-DETR page recommending `deterministic=False`.

```python
from ultralytics import RTDETR

RTDETR("rtdetr-l.pt").train(data="coco8.yaml", epochs=100, imgsz=640, deterministic=False)
```

### Why RT-DETR cannot be deterministic

`RTDETRDecoder` uses `MSDeformAttn`, which calls `F.grid_sample`. Under `torch.use_deterministic_algorithms(True, warn_only=True)` (what `init_seeds` sets) PyTorch reports:

```
UserWarning: grid_sampler_2d_backward_cuda does not have a deterministic implementation,
but you set 'torch.use_deterministic_algorithms(True, warn_only=True)'.
```

Because `warn_only=True`, the run does not fail — it silently continues with a non-deterministic gradient. Two consequences make this unrecoverable rather than a minor rounding concern:

- `grid_sample` sits in the **decoder**, so its gradient flows back through the encoder and the entire backbone. Every parameter is downstream of the non-deterministic kernel.
- `RTDETRDetectionLoss` uses Hungarian bipartite matching, a **discrete** argmin. A 1-ULP gradient difference flips a query-to-target assignment, so divergence is amplified discontinuously rather than staying at rounding scale.

Only the CUDA backward is affected; the CPU backward is deterministic, and forward-only paths (`val`, `predict`, `export`) are unaffected.

### Measured: determinism is not achieved

`rtdetr-l`, fp32 (no AMP), identical seed, identical batch, 5 SGD steps, `deterministic=True`, same GPU. SHA-256 over the full `state_dict`:

| run | weights sha256 | loss @ 5 steps |
| --- | --- | --- |
| rtdetr-l #1 | `35bb7c6fd0e28d24…` | 62.3840179443 |
| rtdetr-l #2 | `4ae8a13ac67428ec…` | 63.0046463013 |
| rtdetr-l #3 | `dd7317231a4d2576…` | 62.4183883667 |
| rtdetr-l #4 | `bd37c9367734d4db…` | 62.1620864868 |
| rtdetr-l #5 | `bca2a03a030c64e2…` | 61.8274497986 |
| yolo11n #1-3 (control) | `f52df496687d852b…` x3 | 46.3437728882 x3 |

5/5 unique for RT-DETR; the YOLO11n control is bitwise identical across 3 separate processes, confirming the harness itself is sound.

### Measured: partial determinism does not reduce divergence

`deterministic=True` still makes every other op deterministic, so does it at least shrink the divergence? No. 12 runs per setting, 66 pairwise weight comparisons each:

| metric | `deterministic=True` | `deterministic=False` |
| --- | --- | --- |
| final-loss sd | 0.364 | 0.456 |
| final-loss range | 1.413 | **1.382** |
| mean pairwise max abs weight delta | 2.998e-01 | 3.075e-01 |
| mean pairwise relative L2 | 3.693e-02 | 3.780e-02 |
| parameters differing between runs | **99.79%** | **99.79%** |

The weight-space metrics are indistinguishable and the loss range is marginally *worse* with `deterministic=True`. F-test on the loss variances: `F = 1.564`, df = (11, 11), two-sided **p = 0.47**.

At n=5 this looked like a real effect (loss range 0.474 vs 1.030, a 2.2x gap) and it inverted once n reached 12 — the signature of a small-sample draw, not an effect.

### Measured: the cost

`rtdetr-l`, bs=8, imgsz=640, AMP, RTX PRO 6000 Blackwell, torch 2.11.0+cu128. Median ms/iter over 25 timed iterations after 8 warmup, 3 alternating A/B rounds:

| round | `deterministic=True` | `deterministic=False` |
| --- | --- | --- |
| 1 | 146.9 ms | 120.7 ms |
| 2 | 160.3 ms | 117.7 ms |
| 3 | 161.1 ms | 120.2 ms |

Roughly **1.3x slower**, about 25% of training throughput. Decomposing which of the three settings `init_seeds` applies is responsible (min ms/iter, least-contended sample):

| setting | min ms/iter | cost |
| --- | --- | --- |
| none | 94-95 | baseline |
| `cudnn.deterministic = True` | 94-97 | free |
| `CUBLAS_WORKSPACE_CONFIG=:4096:8` | 98-103 | ~6% |
| `use_deterministic_algorithms(True, warn_only=True)` | 111 | ~17% |
| all three | 116-122 | ~25% |

The dominant cost is `use_deterministic_algorithms`, which forces deterministic gather/scatter/`index_put` kernels — heavily used by RT-DETR's decoder, denoising groups, and matcher. It is also the setting whose entire purpose `grid_sample` defeats.

### Scope

Documentation only, no behavior change. `RTDETRTrainer` still honors `deterministic` as passed; users opt into the speed themselves.

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

### 🌟 Summary
Updated the RT-DETR documentation to recommend `deterministic=False` for CUDA training because `F.grid_sample` prevents deterministic backward execution while deterministic settings reduce throughput.

### 📊 Key Changes
- Added a **Deterministic Training** tip to `docs/en/models/rtdetr.md`.
- Documents using `deterministic=False` with RT-DETR training on CUDA and PyTorch 2.0 or later.
- Explains that deformable attention uses `F.grid_sample`, which lacks a deterministic CUDA backward implementation.
- Notes that `deterministic=True` may reduce training throughput without making CUDA training reproducible.
- Clarifies that `seed` still controls weight initialization, data order, and augmentation sampling.

### 🎯 Purpose & Impact
- RT-DETR users are guided to opt out of deterministic algorithms during CUDA training to avoid unnecessary performance costs.
- No user-facing code behavior change; `RTDETRTrainer` continues to honor the `deterministic` argument provided by the user.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @artest08, thank you for submitting a `ultralytics/ultralytics` 🚀 PR! This automated message confirms your contribution was received, and an Ultralytics engineer will assist with the review. To ensure a seamless integration of your work, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_ — Bruce Lee

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). Don't hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! 🚀

File: docs/en/models/rtdetr.md [hunk 1]
```diff
@@ -77,6 +77,10 @@ This example provides simple RT-DETR training and inference examples. For full d
         yolo predict model=rtdetr-l.pt source=path/to/bus.jpg
         ```
 
+!!! tip "Deterministic Training"
+
+    Set `deterministic=False` when training RT-DETR. The `F.grid_sample` call in its deformable attention has no deterministic CUDA backward, so `deterministic=True` cannot make a run reproducible yet still costs around 25% of training throughput.
```
Comment (UltralyticsAssistant) [inline docs/en/models/rtdetr.md]: 📝 **LOW**: Scope this recommendation to CUDA training, and preferably PyTorch 2.0+ where `init_seeds` actually enables deterministic algorithms. The trainer also supports CPU, where this `grid_sample` limitation does not apply; as written, a CPU user following the tip unnecessarily disables reproducibility. PyTorch documents the limitation as CUDA-backward-specific. 

**Suggested change:**
```suggestion
    On CUDA with PyTorch 2.0+, set `deterministic=False` when training RT-DETR. Its deformable attention uses `F.grid_sample`, which has no deterministic CUDA backward, so `deterministic=True` cannot make a CUDA run reproducible yet still costs around 25% of training throughput on the tested setup.
```

**Review by UltralyticsAssistant — COMMENTED** [review]
## 🔍 PR Review

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)</sub>

Reviewed the complete 4-line documentation diff. The recommendation is directionally correct for CUDA, but it is overbroad for CPU and older supported PyTorch versions; qualify the tip before merging.

<details><summary>💬 Posted 1 inline comment</summary>

- 📝 **LOW** `docs/en/models/rtdetr.md:82` Scope this recommendation to CUDA training, and preferably PyTorch 2.0+ where `init_seeds` actually enables deterministic algorithms. The trainer also supports CPU, where this `grid_sample` limitation does not apply; as written, a CPU user following the tip unnecessarily disables reproducibility. PyTorch documents the limitation as CUDA-backward-specific.
</details>

**Review by glenn-jocher — APPROVED** [review]

**Comment by UltralyticsAssistant** [comment]
🎉 [PR #25965](https://github.com/ultralytics/ultralytics/pull/25965) has been merged—thank you @artest08, with contributions from @glenn-jocher!

> “The details are not the details. They make the design.” — Charles Eames

This update adds clear RT-DETR guidance to use `deterministic=False` for CUDA training with PyTorch 2.0+ because `F.grid_sample` lacks a deterministic CUDA backward implementation. It also clarifies the throughput trade-off, reproducibility limitations, and the continued role of `seed` in initialization, data ordering, and augmentation sampling.
